### ISSUE TITLE
Shouldn't exact matches for headings be weighed highest?

### ISSUE THREAD

**Issue by hirasso** [CloudCannon/pagefind#437]
Hi there, 

we just updated pagefind to 1.0.2 for [the docs for swup](https://swup.js.org/) and it's amazing! Thanks for all your work on this project.

Playing around with it, I noticed something that might or might not be possible to generalize: When I search for "Plugins", I'm getting the following results:

![image](https://github.com/CloudCannon/pagefind/assets/869813/c12a9f7d-4c9a-4347-ab3c-6a5b1de0c136)

Intuitively I'd think that the page that *exactly* matches the search term ("Plugins", highlighted in the screenshot results above) should be rated highest. Sure it's possible for us to manually give the "plugins" page a very high rating – but maybe there is a way for pagefind to get smart enough to return pages with an exact match in the main heading first? 

I'd be happy to hear your opinion on that before we start implementing manual ranking.

**Comment by bglw** [comment]
Ah! Interesting.

So first for reasoning on why the results are in that order: headings are given _very strong_ priority — though `/api/properties/` contains an `h2` element of `plugins`, which is also ranked very favorably, causing it to appear so highly in the rankings. 
The other reason is that the page is so short — term frequency plays a big role in ranking, so a short page with multiple matching words will rank better than a long page with a lower density of words.

> Intuitively I'd think that a page who's main heading (h1) exactly matches the search term should be rated highest

On an intuitive level I agree! The trouble is that currently, with the data Pagefind has on hand at the time of ranking, this isn't known. By the time that level of data is loaded into the front end, the rankings are locked in.

When ranking, all we see for these results is:
```js
// Page A word match locations
[{
  "weight": 6,
  "location": 23
}, {
  "weight": 1,
  "location": 27
}]

// Page B word match locations
[{
  "weight": 7,
  "location": 0
}, {
  "weight": 1,
  "location": 12
}, {
  "weight": 1,
  "location": 21
},
/* -- more -- */
]
```

That `weight: 7, location: 0` word is the h1, but we don't know that it's the _only_ word in the h1.

***

I'm sure there's a creative solution here, but nothing immediately comes to mind.

One option would just be to bump `h1` elements default weighting across the board to compensate, but I'd be wary of that impacting other sites in the wrong direction, if they're currently ranking well.

I do want to expose a new configuration option for mapping element selectors to custom weightings, meaning the default h1 rank could be a per-site implementation if people want to tailor their results. But this doesn't solve the "making Pagefind smart enough" goal of doing better by default.

Another option is finding some way to opt h1 elements out of the term frequency penalty — but I'll need to think on that further.

***

Sorry for the essay! Since I have no immediate bright ideas, I think implementing manual ranking is a good first step.

**Comment by hirasso** [comment]
No need to apologize, I love the essay! 😄 ...very interesting to get to know more about pagefind's internals. 

From my experience with implementing search, it's very possible that too much built-in assumptions ("smartness") could hurt more then help. Maybe even the assumption that a perfect match in the `h1` should be ranked highest doesn't hold true. What about if we change the title of the page to "Plugin Ecosystem" or "Plugins Overview" sometime in the future? This would immediately break the ranking again.

From other search engines I know the concept of "pinning" pages to the very top for specific search terms. Maybe that could be a feature idea for pagefind, as well. 

Something like `data-pagefind-pin="plugin,plugins"` or even with regex support: `data-pagefind-pin="plugins?"` could *pin* the page "Plugins" to the very top if users would search for "plugin" or "plugins".

It could also be a meta tag in the `<head>` of the page, like e.g.:

```html
<!-- pin the page to the top when searching for "plugin" or "plugins": -->
<meta name="pagefind-pin" content="plugin,plugins">
```

Just an idea without knowing if this could be feasible with the architecture of pagefind 🙂

**Comment by daun** [comment]
Given the implementation details kindly explained by @bglw, I'd opt for manual ranking here. Intuitively, for most sites, ranking pages the highest that have an exact match between search term and h1/title makes sense, but if that's not encoded in the ranking data, ranking them during display should work as well. I can imagine treating the h1 as a special `title` property of a page apart from the other contents, as that's how they're handled in display as well. But that's probably opening a whole different can of worms...

**Comment by bglw** [comment]
Good directions to think about, thanks to both of ya 🙏

I need to brush up CloudCannon's documentation and search, so that might prove a good place to experiment with some pinning on a site I'm familiar with. Will update here if I land on anything!

**Comment by bglw** [comment]
Some new thoughts on this; I think I'm going to try get metadata into the index in a way that it can be queried as part of a search. This would allow you to do a freeform search for the word `plugins`, or run a search specifically for title metadata containing the word `plugin`, or some combination of the two.

Still more ideation needed, but I think having that data combined with exposing more configuration on how it is counted will allow people to tailor search to their site content.

**Comment by clydebarrow** [comment]
My request is not to exactly match titles, but filenames:

It would be nice IMHO if a search that exactly matched the name of a page would deliver that page as the top hit. E.g. searching for "I2C" would put a page called "I2C.html" at the top (without case sensitivity of course.)

Are filenames even used at present?

**Comment by bglw** [comment]
Filenames are currently unused apart from building URLs, but the URLs are present in the result fragments, so what you're after would still be solved by #532, as it is the precursor to matching files based on any "non-content" fields

**Comment by quadratecode** [comment]
I agree that matches in page titles are not adquately weighed. After tinkering with [all available parameters](https://pagefind.app/docs/ranking/) I could not get exact matches in the title tag with `data-pagefind-weight="10"` to show up on the top of the results list. Here is an example:

![image](https://github.com/CloudCannon/pagefind/assets/47365835/a9310a94-8abb-4202-b7fc-a34ead862552)

There are also some other rankings which I do not understand - e.g. here:

![image](https://github.com/CloudCannon/pagefind/assets/47365835/bafc4def-2457-4040-9f3f-2cde12ca8271)

In the above image, the first result has 6 matches while the second one has 49 matches. Changing the parameters did not seem to do much.

If you want, you can try it yourself here: [https://www.zhlaw.ch/](https://www.zhlaw.ch/)

Anyways - just also wanted to say: Amazing project, thank you so much for your work and offering this under a permissible license! 🙏

**Comment by sonderant** [comment]
I agree that the exact match in the title should be given more priority. For example, in my site, I have some articles (360) that begin with the word and number, e.g. Article 1, Article 206, etc. When searching with Pagefind, any article between 1 and 30 does not show up, even in the top ten. For example, "Article 1" (no quotes) returns all articles in the 100-199 range, or articles with the word "article" or the number 1 in them. But any article past number 36 shows up in the top ten.

**Comment by paulmoloneyr3** [comment]
Hi there, just to add my voice here. I've implemented Pagefind search on a Hugo docs site, and find that - for example - if I search for "Release Notes", the actual topic with the title "Release Notes" comes below other pages that refer to it. I would love to know how to have topics titled "Release Notes" at the top in these instances.

**Comment by quadratecode** [comment]
I revisited this issue with [1.1.1](https://github.com/CloudCannon/pagefind/releases/tag/v1.1.1) and after tweaking all available parameters it seems to persists for me. I observed the following behaviour:

- Even if I assigned `data-pagefind-weight="10"` to titles, results with an exact match in the title would rank below results without an exact match. I suspect the cause to be `termSimilarity` which cannot be turned off completely. On very short queries (e.g. three letters), this can lead to unexpected results. One example would be if I searched for "IDG" (without quotes, see comment above), the result would rank second with three exact occurences (both weighted 10, one weighted 7) against the second result with 32 exact occurences (all weighted 1) and 3 similar occurences (2 weighed 10 and 1 weighed 7). A possible solution could maybe be to further enhance the weight of exact matches.
- If I search "IDG" (with quotes) the result with three exact occurences (both weighted 10, one weighted 7) drops to the bottom of results, below any results with a higher count regardless of weight. A possible solution could maybe be to also consider weights when searching for exact matches.

**Comment by bglw** [comment]
Shipped in [v1.5.0](https://github.com/Pagefind/pagefind/releases/tag/v1.5.0) 🙂 This is now way way way way better
